File: cpp/problems/leetcode/0127_word_ladder/solution.cpp
// LeetCode 127: Word Ladder
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> words(wordList.begin(), wordList.end());
        if (!words.count(endWord)) return 0;
        queue<pair<string, int>> q;
        q.push({beginWord, 1});
        unordered_set<string> seen = {beginWord};
        while (!q.empty()) {
            auto [w, depth] = q.front();
            q.pop();
            if (w == endWord) return depth;
            for (int i = 0; i < (int)w.size(); ++i) {
                for (char c = 'a'; c <= 'z'; ++c) {
                    string nw = w;
                    nw[i] = c;
                    if (words.count(nw) && !seen.count(nw)) {
                        seen.insert(nw);
                        q.push({nw, depth + 1});
                    }
                }
            }
        }
        return 0;
    }
};

int main() {
    vector<string> wl{"hot", "dot", "dog", "lot", "log", "cog"};
    assert(Solution().ladderLength("hit", "cog", wl) == 5);
    vector<string> wl2{"hot", "dot", "dog"};
    assert(Solution().ladderLength("hit", "cog", wl2) == 0);
    vector<string> wl3{"a", "b", "c"};
    assert(Solution().ladderLength("a", "c", wl3) == 2);
    cout << "0127 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0127_word_ladder/solution.py
"""LeetCode 127: Word Ladder."""

from __future__ import annotations

from collections import deque


class Solution:
    def ladderLength(
        self,
        beginWord: str,
        endWord: str,
        wordList: list[str],
    ) -> int:
        words = set(wordList)
        if endWord not in words:
            return 0
        q: deque[tuple[str, int]] = deque([(beginWord, 1)])
        seen = {beginWord}
        alpha = "abcdefghijklmnopqrstuvwxyz"
        while q:
            w, depth = q.popleft()
            if w == endWord:
                return depth
            for i in range(len(w)):
                for c in alpha:
                    nw = w[:i] + c + w[i + 1 :]
                    if nw in words and nw not in seen:
                        seen.add(nw)
                        q.append((nw, depth + 1))
        return 0


if __name__ == "__main__":
    wl = ["hot", "dot", "dog", "lot", "log", "cog"]
    assert Solution().ladderLength("hit", "cog", wl) == 5
    assert Solution().ladderLength("hit", "cog", ["hot", "dot", "dog"]) == 0
    assert Solution().ladderLength("a", "c", ["a", "b", "c"]) == 2
    print("0127 OK")
